### PR TITLE
Added a new tool and updated an existing tool

### PR DESCRIPTION
**Pull request by qgao-hd**
1, updated tool "image_registration_affine": more options of the image similarity metric; support of RGB(A) images

2, added new tool "3d_tensor_feature_dimension_reduction"

### CODE REVIEW

**Comment by qgao-hd** [comment]
It is weird that the .npz file (compressed array; the only input) cannot be correctly loaded.


Standard Error:
Traceback (most recent call last):
  File "/home/runner/.planemo/planemo_tmp_7m7wh7ts/3d_tensor_feature_dimension_reduction.py", line 39, in <module>
    feature_dimension_reduction(args.tensor_fn, args.tiff_fn, args.nCh)
  File "/home/runner/.planemo/planemo_tmp_7m7wh7ts/3d_tensor_feature_dimension_reduction.py", line 19, in feature_dimension_reduction
    assert isinstance(npobj, np.lib.npyio.NpzFile), 'input must be a npz file'
AssertionError: input must be a npz file

**Comment by bgruening** [comment]
lets see if this works, it can be that numpy __needs__ a proper fileending.

**Comment by qgao-hd** [comment]
> It does not. Maybe your test file is currupted?

I downloaded the test file from my fork and made a test.

It can be loaded and processed on my local machine without any problem.

**Comment by qgao-hd** [comment]
I am using Python 3.7.10 on my local machine (MacBook Air M1 chip). I have Planemo installed. But 'planemo test' never works properly ... Don't know why.

I will install Planemo on my Linux PC in the office and try planemo test again on Monday.

**Comment by qgao-hd** [comment]
> Does you local planemo also fail?

I am using Python 3.7.10 on my local machine (MacBook Air M1 chip). I have Planemo installed. But 'planemo test' never works properly ... Don't know why.

I will install Planemo on my Linux PC in the office and try planemo test again on Monday.

**Comment by bgruening** [comment]
Thanks! No clue how the new OSX does that. Conda needs to work or Docker, maybe that is also the problem?

**Comment by qgao-hd** [comment]
> Thanks! No clue how the new OSX does that. Conda needs to work or Docker, maybe that is also the problem?

I tried planemo test on my Linux PC. Unfortunately I always had the same error as on my Mac "module not found" for all tools with Python script. It seemed some dependency problem of Python packages. But I have no clue at all.

**Comment by bgruening** [comment]
> > Thanks! No clue how the new OSX does that. Conda needs to work or Docker, maybe that is also the problem?
> 
> I tried planemo test on my Linux PC. Unfortunately I always had the same error as on my Mac "module not found" for all tools with Python script. It seemed some dependency problem of Python packages. But I have no clue at all.

So its also failing for you on Linux?

You can try with docker `planemo test --biocontainers`

**Comment by qgao-hd** [comment]
> It does not. Maybe your test file is currupted?

I changed the format of input to hdf5 and it works. 

But now the error is:
Output fn_out:  different than expected, difference (using diff):
( /tmp/tmp18dfrl6atensor_r.tif v. /tmp/tmpnjj4fs9ptensor_r.tif )
Binary data detected, not displaying diff

As the output is multi-channel 32bit tiff, I do not know whether the tiff output is not properly saved or Galaxy does not support this tiff format.

File: tools/3d_tensor_feature_dimension_reduction/3d_tensor_feature_dimension_reduction.xml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        <param name="nCh" type="integer" value="5" optional="true" min="1" max="10" label="Reduced dimensions of features" />
+    </inputs>
+    <outputs>
+        <data format="tiff" name="fn_out" from_work_dir="output.tif" />
+    </outputs>
+    <tests>
+        <test>
+            <param name="fn_in" value="tensor.h5"/>
+            <param name="nCh" value="3"/>
+            <output name="fn_out" value="tensor_r.tif" ftype="tiff" />
```
Comment (bgruening) [inline tools/3d_tensor_feature_dimension_reduction/3d_tensor_feature_dimension_reduction.xml:28]: by default galaxy will diff your files, this is not possible with binary data. You can use for example this here: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output-assert-contents or any otherway to test that your tool is producing what you expect :)

**Comment by qgao-hd** [comment]
> So its also failing for you on Linux?
> 
> You can try with docker `planemo test --biocontainers`

Local Planemo (planemo test) fails on both my Linux and Mac with ModuleNotFoundError.

planemo test --biocontainers does not help.

**Comment by qgao-hd** [comment]
> So its also failing for you on Linux?
> 
> You can try with docker `planemo test --biocontainers`

I have miniconda3 with Python3.7 installed. This is how I installed Planemo and run planemo test:
$ virtualenv .venv; . .venv/bin/activate 
$ pip install planemo (my pip=21.1.1)
$ planemo lint (this works)
$ planemo test (this fails when there is Python script)
$ planemo test --biocontainers (also fails)

**Comment by qgao-hd** [comment]
The output was not correctly saved.

Problems:
Output fn_out:  different than expected
Expected file size was 170000, actual file size was 0 (difference of 20000 accepted)

File: tools/3d_tensor_feature_dimension_reduction/3d_tensor_feature_dimension_reduction.xml (hunk 1 above)
Comment (qgao-hd) [reply tools/3d_tensor_feature_dimension_reduction/3d_tensor_feature_dimension_reduction.xml]: Now I check the output file size

**Comment by bgruening** [comment]
That means that no output file is created, or 0 bytes one. So maybe you are really missing a dependency?

I can only look at this, this evening.

**Comment by qgao-hd** [comment]
> That means that no output file is created, or 0 bytes one. So maybe you are really missing a dependency?
> 
> I can only look at this, this evening.


Finally all checks passed!

It turned out that file extension should not be changed/updated in Python script like this:
tifffile.imwrite(tiff_fn + '.tif', ...)

**Pull request merged**